File: trunk/MyArrayController.m
//
//  MyArrayController.m
//  ManiaDB
//
//  Created by Appledelhi on 11/30/08.
//  Copyright 2008 Appledelhi. All rights reserved.
//

#import "MyArrayController.h"

static NSString* MyTableViewDataType = @"MyTableViewDataType";

@implementation MyArrayController

- (BOOL)tableView:(NSTableView*)tableView writeRowsWithIndexes:(NSIndexSet*)rowIndexes
     toPasteboard:(NSPasteboard*)pboard
{
  NSData* data = [NSKeyedArchiver archivedDataWithRootObject:rowIndexes];
  [pboard declareTypes:[NSArray arrayWithObject:MyTableViewDataType] owner:self];
  [pboard setData:data forType:MyTableViewDataType];
  return YES;
}

- (NSDragOperation)tableView:(NSTableView*)tableView validateDrop:(id <NSDraggingInfo>)info
                 proposedRow:(NSInteger)row proposedDropOperation:(NSTableViewDropOperation)dropOperation
{
  [tableView setDropRow:row dropOperation:NSTableViewDropAbove];
  return NSDragOperationMove;
}

- (BOOL)tableView:(NSTableView*)tableView acceptDrop:(id <NSDraggingInfo>)info row:(NSInteger)row
    dropOperation:(NSTableViewDropOperation)dropOperation
{
  NSPasteboard* pboard = [info draggingPasteboard];
  NSData* rowData = [pboard dataForType:MyTableViewDataType];
  NSIndexSet* rowIndexes = [NSKeyedUnarchiver unarchiveObjectWithData:rowData];
  int dragRow = [rowIndexes firstIndex];
  id obj = [[self arrangedObjects] objectAtIndex:dragRow];
  [self insertObject:obj atArrangedObjectIndex:row];
  [self removeObjectAtArrangedObjectIndex:dragRow > row ? dragRow + 1 : dragRow];
  return YES;
}

- (void)awakeFromNib
{
  [_tableView setDraggingSourceOperationMask:NSDragOperationMove forLocal:YES];
  [_tableView registerForDraggedTypes:[NSArray arrayWithObject:MyTableViewDataType]];
}

@end

File: trunk/MyArrayController.rb
MyTableViewDataType = "MyTableViewDataType"

class MyArrayController < NSArrayController
  attr_accessor :tableView
  
  def tableView_writeRowsWithIndexes_toPasteboard(tableView, rowIndexes, pboard)
    data = NSKeyedArchiver.archivedDataWithRootObject(rowIndexes)
    pboard.declareTypes_owner([MyTableViewDataType], self)
    pboard.setData_forType(data, MyTableViewDataType)
    true
  end
  
  def tableView_validateDrop_proposedRow_proposedDropOperation(tableView, info, row, op)
    tableView.setDropRow_dropOperation(row, NSTableViewDropAbove)
    NSDragOperationMove
  end
  
  def tableView_acceptDrop_row_dropOperation(tableView, info, row, op)
    pboard = info.draggingPasteboard
    rowData = pboard.dataForType(MyTableViewDataType)
    rowIndexes = NSKeyedUnarchiver.unarchiveObjectWithData(rowData)
    dragRow = rowIndexes.firstIndex
    obj = arrangedObjects.objectAtIndex(dragRow)
    insertObject_atArrangedObjectIndex(obj, row)
    removeObjectAtArrangedObjectIndex(dragRow > row ? dragRow + 1 : dragRow)
    true
  end
  
  def awakeFromNib
    @tableView.setDraggingSourceOperationMask_forLocal(NSDragOperationMove, true)
    @tableView.registerForDraggedTypes([MyTableViewDataType])
  end
end
